### DISCUSSION TITLE
CLI Thoughts?

### DISCUSSION THREAD

**Discussion by KennanHunter** [rolldown/rolldown#592] [General]
I decided it would be best to pose some open-ended questions before I or anyone else started on a Command Line Interface for rolldown. 

What language do we go with?
- If we go with JS/TS there's a lot of decent tooling for making particularly aesthetic CLIs. The biggest pros for doing it in JS/TS would be that bundling with npm is super easy, and overall development would be a bit faster. The cons here are that startup time would be much worse, and we wouldn't be leveraging the convenience that comes from having Rust end-to-end. [This is how SWC does it](https://github.com/swc-project/pkgs/tree/main/packages/cli). Another benefit of having it written in JS/TS is making sure our javascript API is solid.
- If we go with Rust we have to solve the bundling issue of running binary Rust on npm, a problem that has already been solved by other projects using npm's package.json [bin parameter](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#bin), which is used by [ESBuild](https://github.com/evanw/esbuild/blob/9d1777f23d9b64c186345223d92f319e59388d8b/npm/esbuild/package.json#L17-L19), [Biome](https://github.com/biomejs/biome/blob/55257107d7ac8160eb26397ad902eb42bee1799e/packages/%40biomejs/biome/package.json#L4), oxc, and certainly more.

Do we try to match pre-existing projects?
- If we really tried we could probably be a drop-in replacement for rollup, but that would lock us into a pre-existing system.
- We could just try to keep the general feel similar, much like how npm and yarn feel mostly the same.
- We could also just go rogue and decide to build the tooling up from scratch built around our needs, particularly as Vite compatibility is the main goal here, and less so viability as an individual tool.

We don't need to know these answers fully right now but I think these are good starting questions. Personally, I'm leaning towards writing it in Rust and while It'd be more fun to do the tooling from scratch, I think it makes sense to at least preserve the feel of rollup.

**Comment by overlookmotel** [comment]
One question if going with Rust end-to-end is how to handle plugins written in JS. Presumably this would require embedding a JS runtime inside Rolldown. This would be possible, but no small task.

**Comment by kazupon** [comment]
I think the CLI Which language to implement depends on the requirements you have for the CLI.

If the CLI is performance critical, I would prefer to implement it in Rust; if it provides simple bundle functionality like the rollup cli, I think a JS/TS implementation is sufficient.

If we will implement it in Rust, we have to figure out how to make the plugin work. rollup plugin is implemented in Node's API, and as already mentioned, it is very difficult to embed a JS runtime equivalent to Node.

Since Vite also has a CLI, I think the first thing to consider is what role the rollodown CLI will have.

**Reply by kazupon** [reply]
BTW, I've just started a personal side project. :)
https://github.com/kazupon/rolldown-cli

**Comment by yyx990803** [comment]
My current thoughts:

1. The CLI should be in JS/TS. CLI startup time not so much of a concern as most users would run it via npm scripts anyway - faster iteration and easier builds separate from the Rust core should be worth it.

2. Behavior wise, I'd say it will be "same as Rollup CLI to a reasonable extent", and we can enhance / add additional features where it makes sense. Kinda like your Yarn to npm analogy.

**Reply by hyfdev** [reply]
+1. Many explorations could be done, but for a practical and current implementation:

- The CLI should be written in JS/TS, since we need to support JS plugins.
- Rollup's API is elegant in most of parts. Aligning with it is practical. This doesn't mean we will be limited, we could still extend it if there are something unreasonable.

---

I personally would hope this cli is written in js with jsdoc. Having another build step on cli isn't easy or hard, but if we could avoid it, I tend to avoid it.
